express-api (Tara)

trangtara opened the pull request.

KarolCislowski commented on README.md:
```diff
@@ -1,13 +1,12 @@
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+https://goodbookscollection.herokuapp.com/books
```
I tried to open your app, but Heroku returns information that this app is not created.

trangtara replied:
Hey Karol, the link to Heroku is restored. Please check it out. Thank you: 
https://goodbookscollection.herokuapp.com/books

KarolCislowski commented on server.js:
```diff
@@ -1,32 +1,150 @@
 
+app.get('/books', (req, res) => {
+  const books = booksData
+  const page = req.query.page ? req.query.page -1 : 0 //page 0 as default, page number is -1 compared to the array index
+    const pageSize = req.query.pageSize ?? 20 //default is 20
+    const startIndex = page * pageSize
+    const endIndex = startIndex + pageSize
+    const bookPerPage = books.slice(startIndex, endIndex)
+    const returnObj = {
+      amountOfData: books.length,
+      amounfOfPages: Math.ceil(books.length / pageSize),
+      currentPage: page + 1,
+      pageSize: pageSize,
+      startIndex: startIndex,
+      endIndex: endIndex,
+      results: bookPerPage
+    }
```
Nice and clean pagination code 👍🏻  Maybe it would be good to extract it to the separate function to gain the ability to reuse it?

KarolCislowski commented on server.js:
```diff
@@ -1,32 +1,150 @@
+    }
+
+    if (bookPerPage.length === 0) {
+      res.status(404).json(ERROR_DATA_NOT_FOUND)
```
nice error handling

KarolCislowski commented on server.js:
```diff
@@ -1,32 +1,150 @@
+})
+
+//sort books by average_rating
+app.get('/books/rating/sort', (req, res) => {
+  const { sort } = req.query
+  const sortbyRating = booksData.sort((a,b) => {
+    if (sort === 'asc') {
+      return +a.average_rating - +b.average_rating
+    } else if ( sort === 'des') {
+      return +b.average_rating - a.average_rating
+    } 
+  })
+
+  if (sortbyRating.length === 0) {
+    return res.status(404).json(ERROR_DATA_NOT_FOUND)
+  } else {
+    res.json(sortbyRating)
+  }
+})
+
+//sort books by number of pages
+app.get('/books/pages/sort', (req, res) => {
+  const { sort } = req.query
+  const sortbyPages = booksData.sort((a,b) => {
+    if (sort === 'asc') {
+      return +a.num_pages - +b.num_pages
+    } else if ( sort === 'des') {
+      return +b.num_pages - +a.num_pages
+    }
+  })
+
+  if (sortbyPages.length === 0) {
+    return res.status(404).json(ERROR_DATA_NOT_FOUND)
+  } else {
+    res.json(sortbyPages)
+  }
+})
+
+
```
You created a nice sorting code here, but you repeated it. It is a big sign that you should create a separate sorting function to call anytime it is needed 🤓

KarolCislowski commented on server.js:
```diff
@@ -1,32 +1,150 @@
+})
+
+//sort books by average_rating
+app.get('/books/rating/sort', (req, res) => {
```
maybe creating `/books/sort/:field` would make you create just one endpoint with the possibility to sort it as your user wants?

KarolCislowski reviewed (changes requested):
The code looks great! 🦄  Please check what happened with your hosted on Heroku version, to make testing possible 🤓

trangtara commented:
Oh, i guess i deleted the collection on compass. I'll restore it. Thanks!

On Wed, 27 Jan 2021 at 11:35 Karol Cislowski <notifications@github.com>
wrote:

> *@KarolCislowski* requested changes on this pull request.
>
> The code looks great! 🦄 Please check what happened with your hosted on
> Heroku version, to make testing possible 🤓
> ------------------------------
>
> In README.md
> <https://github.com/Technigo/project-express-api/pull/135#discussion_r565193757>
> :
>
> >
>
>  ## View it live
>
>
>
> -Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
>
> +https://goodbookscollection.herokuapp.com/books
>
> I tried to open your app, but Heroku returns information that this app is
> not created.
> ------------------------------
>
> In server.js
> <https://github.com/Technigo/project-express-api/pull/135#discussion_r565194962>
> :
>
> > +  const page = req.query.page ? req.query.page -1 : 0 //page 0 as default, page number is -1 compared to the array index
>
> +    const pageSize = req.query.pageSize ?? 20 //default is 20
>
> +    const startIndex = page * pageSize
>
> +    const endIndex = startIndex + pageSize
>
> +    const bookPerPage = books.slice(startIndex, endIndex)
>
> +    const returnObj = {
>
> +      amountOfData: books.length,
>
> +      amounfOfPages: Math.ceil(books.length / pageSize),
>
> +      currentPage: page + 1,
>
> +      pageSize: pageSize,
>
> +      startIndex: startIndex,
>
> +      endIndex: endIndex,
>
> +      results: bookPerPage
>
> +    }
>
>
> Nice and clean pagination code 👍🏻 Maybe it would be good to extract it
> to the separate function to gain the ability to reuse it?
> ------------------------------
>
> In server.js
> <https://github.com/Technigo/project-express-api/pull/135#discussion_r565195189>
> :
>
> > +    const pageSize = req.query.pageSize ?? 20 //default is 20
>
> +    const startIndex = page * pageSize
>
> +    const endIndex = startIndex + pageSize
>
> +    const bookPerPage = books.slice(startIndex, endIndex)
>
> +    const returnObj = {
>
> +      amountOfData: books.length,
>
> +      amounfOfPages: Math.ceil(books.length / pageSize),
>
> +      currentPage: page + 1,
>
> +      pageSize: pageSize,
>
> +      startIndex: startIndex,
>
> +      endIndex: endIndex,
>
> +      results: bookPerPage
>
> +    }
>
> +
>
> +    if (bookPerPage.length === 0) {
>
> +      res.status(404).json(ERROR_DATA_NOT_FOUND)
>
>
> nice error handling
> ------------------------------
>
> In server.js
> <https://github.com/Technigo/project-express-api/pull/135#discussion_r565197318>
> :
>
> > +app.get('/books/rating/sort', (req, res) => {
>
> +  const { sort } = req.query
>
> +  const sortbyRating = booksData.sort((a,b) => {
>
> +    if (sort === 'asc') {
>
> +      return +a.average_rating - +b.average_rating
>
> +    } else if ( sort === 'des') {
>
> +      return +b.average_rating - a.average_rating
>
> +    }
>
> +  })
>
> +
>
> +  if (sortbyRating.length === 0) {
>
> +    return res.status(404).json(ERROR_DATA_NOT_FOUND)
>
> +  } else {
>
> +    res.json(sortbyRating)
>
> +  }
>
> +})
>
> +
>
> +//sort books by number of pages
>
> +app.get('/books/pages/sort', (req, res) => {
>
> +  const { sort } = req.query
>
> +  const sortbyPages = booksData.sort((a,b) => {
>
> +    if (sort === 'asc') {
>
> +      return +a.num_pages - +b.num_pages
>
> +    } else if ( sort === 'des') {
>
> +      return +b.num_pages - +a.num_pages
>
> +    }
>
> +  })
>
> +
>
> +  if (sortbyPages.length === 0) {
>
> +    return res.status(404).json(ERROR_DATA_NOT_FOUND)
>
> +  } else {
>
> +    res.json(sortbyPages)
>
> +  }
>
> +})
>
> +
>
> +
>
>
> You created a nice sorting code here, but you repeated it. It is a big
> sign that you should create a separate sorting function to call anytime it
> is needed 🤓
> ------------------------------
>
> In server.js
> <https://github.com/Technigo/project-express-api/pull/135#discussion_r565198437>
> :
>
> > +    const lowercaseLang = langString.toLowerCase()
>
> +    const splitLangData = lowercaseLang.split('')
>
> +    const sliceLangData = splitLangData.slice(0, 2)
>
> +
>
> +    return sliceLangInput[0] === sliceLangData[0] && sliceLangInput[1]=== sliceLangData[1]
>
> +  })
>
> +
>
> +  if (filterByLanguage.length === 0) {
>
> +      return res.status(404).json(ERROR_DATA_NOT_FOUND)
>
> +    } else {
>
> +      res.json(filterByLanguage)
>
> +    }
>
> +})
>
> +
>
> +//sort books by average_rating
>
> +app.get('/books/rating/sort', (req, res) => {
>
>
> maybe creating /books/sort/:field would make you create just one endpoint
> with the possibility to sort it as your user wants?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Technigo/project-express-api/pull/135#pullrequestreview-577188726>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRXH3ISRGTKDW2F5PBLZ2LS37T7FANCNFSM4UZ2BTNQ>
> .
>
-- 
This email is sent from my iPhone. Please accept its short nature and some
typo.

KarolCislowski reviewed (approved):
Now it works perfectly 🥳
